### ISSUE TITLE
Incorrect mypy types

### ISSUE THREAD

**Issue by abrisco** [agronholm/cbor2#216]
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### cbor2 version

5.6.1

### Python version

3.11

### What happened?

Running mypy on my code which uses cbor2 revealed some things which appear to be incorrect types
- https://github.com/agronholm/cbor2/blob/5.6.1/cbor2/_decoder.py#L794 (should be `-> Any`)
- https://github.com/agronholm/cbor2/blob/5.6.1/cbor2/_decoder.py#L825 (should be `: Io[bytes]`)
- https://github.com/agronholm/cbor2/blob/5.6.1/cbor2/_decoder.py#L829 (should be `-> Any`)
- https://github.com/agronholm/cbor2/blob/5.6.1/cbor2/_encoder.py#L751 (should be `: Any`)
- https://github.com/agronholm/cbor2/blob/5.6.1/cbor2/_encoder.py#L752 (should be `: Io[bytes]`)

### How can we reproduce the bug?

Add mypy to CI and I believe it will catch this.

**Comment by abrisco** [comment]
> We already run mypy in strict mode and it did not catch any of this.

Uh-oh. Does mypy run on the tests? Also, wouldn't mypy in strict mode disallow this missing return type? https://github.com/agronholm/cbor2/blob/5.6.1/cbor2/_decoder.py#L73

**Comment by agronholm** [comment]
>Does mypy run on the tests? 

No, and the test suite itself is untyped.

>Also, wouldn't mypy in strict mode disallow this missing return type?

Yes, since the return type of `__init__()` is always `None`.

**Comment by abrisco** [comment]
> No, and the test suite itself is untyped.

This might be why this was missed, Unless there's a caller trying to use a function with a different type I don't think mypy will otherwise tell you the type is wrong.

> Yes, since the return type of `__init__()` is always `None`.

```python
# demo.py
class Foo:
    def __init__(self):
        print("Hallo")
```

```
(.venv) ~/Code (requires ✗) mypy --strict ./demo.py
demo.py:3: error: Function is missing a return type annotation  [no-untyped-def]
demo.py:3: note: Use "-> None" if function does not return a value
Found 1 error in 1 file (checked 1 source file)
```

Does this suggest mypy is not running with strict mode?

**Comment by agronholm** [comment]
You're getting that error because your `__init__()` method has no type annotations at all. This is not the case with `cbor2`.

**Comment by abrisco** [comment]
> You're getting that error because your `__init__()` method has no type annotations at all. This is not the case with `cbor2`.

Is that not the case here? https://github.com/agronholm/cbor2/blob/5.6.1/cbor2/_decoder.py#L73
<img width="710" alt="image" src="https://github.com/agronholm/cbor2/assets/91817010/826bf0ad-192f-49d3-b460-98f35ed320ac">

**Comment by agronholm** [comment]
I count 4 parameters with type annotations.

**Comment by abrisco** [comment]
> I count 4 parameters with type annotations.

But not the return type. Note that the specific mypy error was
```
demo.py:3: error: Function is missing a return type annotation  [no-untyped-def]
```

**Comment by agronholm** [comment]
Try adding an annotated parameter to your `demo.py` and then running mypy against it.

**Comment by abrisco** [comment]
> Try adding an annotated parameter to your `demo.py` and then running mypy against it.

wow, that's ridiculous behavior from mypy

```python
# demo.py
class Foo:
    def __init__(
            self, 
            name: str
        ):
        print(name)
```
```
Success: no issues found in 1 source file
```

To reign back to the original issue though. The types in the original post still appear to be incorrect. I guess I don't understand mypy well enough to understand how this wasn't caught. Simply loading a cbor will trigger the issue.

```python
import cbor2

with open("example.cbor", "rb") as f:
    content = cbor2.load(f)
```

**Comment by agronholm** [comment]
To be precise, `cbor2.load()` is called by `cbor2.loads()`, but the parameter type matches the value passed by `loads()`, so there is no error reported.

**Comment by abrisco** [comment]
```python
import cbor2

with open("example.cbor", "rb") as f:
    content = cbor2.load(f)
```
```
(.venv) ~/Code: mypy --strict demo.py
demo.py:4: error: Argument 1 to "load" has incompatible type "BufferedReader"; expected "BytesIO"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

**Comment by abrisco** [comment]
> `cbor2.load()` is an entry point to the code, so it's not called from within the package itself, only the test suite. And since the test suite is untyped, mypy checks aren't run there.

Are you suggesting `cbor2.load` is a private/internal function?

**Comment by agronholm** [comment]
No, I'm suggesting that within the scope of the code that mypy checks, only `cbor2.loads()` calls that, and with a BytesIO object which is how `fp` is (incorrectly) typed as.

**Comment by abrisco** [comment]
Also, thank you for helping me understand more about mypy 🙏

**Comment by agronholm** [comment]
Would you mind reviewing the PR?
